### PR TITLE
[Slider] [ThumbTrack] Add Font API and Dynamic Type support

### LINKED ISSUE #8646: [Slider] Expose font and adjustsFontForContentSizeCategory for thumbtrack label
This was filed as an internal issue. If you are a Googler, please visit [b/143287831](http://b/143287831) for more details.

---
#### Internal data
- Associated internal bug: [b/143287831](http://b/143287831)

### PR DESCRIPTION
**Pull request by yarneo**
In this PR the main goal is to add Dynamic Type support. As part of this support I am doing the following:

1. Adding a Font API to ThumbTrack and Slider to be able to set the font of the discrete label to a scalable font.
2. Deprecating the old fontSize API (which isn't used internally).
3. Adding adjustsFontForContentSizeCategory APIs to Slider and ThumbTrack.
4. Resizing the ThumbTrack view accordingly based on the font size instead of having it static.

The snapshot tests in place allow me to present the new support for dynamic type and custom fonts.

Closes #8646

### CODE REVIEW

File: components/Slider/src/MDCSlider.h [hunk 1]
```diff
@@ -367,6 +368,17 @@ IB_DESIGNABLE
...
 @property(nonatomic, assign) BOOL shouldEnableHapticsForAllDiscreteValues;
 
+/**
+ The font of the discrete value label.
+
+ This font will come into effect only when @c numberOfDiscreteValues is larger than 0 and when @c
+ shouldDisplayDiscreteValueLabel is
+ @c YES.
+
+ Defaults to [[MDCTypography fontLoader] regularFontOfSize:12].
```
Comment (romoore) [inline components/Slider/src/MDCSlider.h:378]: Tip: MDCTypography is planned for deprecation in the near future.   Perhaps a comment that it is likely to change in the future as a result?

File: components/Slider/tests/snapshot/MDCSliderSnapshotTests.m [hunk 2]
```diff
@@ -211,11 +215,13 @@ - (void)testPreferredFontForAXXXLContentSizeCategory {
         [UITraitCollection traitCollectionWithPreferredContentSizeCategory:
                                UIContentSizeCategoryAccessibilityExtraExtraExtraLarge];
     self.slider.traitCollectionOverride = aXXXLTraitCollection;
+    UIView *valueLabel = [self.slider.thumbTrack valueForKey:@"_valueLabel"];
+    valueLabel.transform = CGAffineTransformIdentity;
```
Comment (romoore) [inline components/Slider/tests/snapshot/MDCSliderSnapshotTests.m:224]: Was this accidentally left in as debugging/testing code? If not, could you leave a comment explaining why the `transform` needs to be set?

File: components/Slider/tests/snapshot/MDCSliderSnapshotTests.m [hunk 3]
```diff
@@ -237,6 +247,8 @@ - (void)testPreferredFontForXSContentSizeCategory {
     UITraitCollection *xsTraitCollection = [UITraitCollection
         traitCollectionWithPreferredContentSizeCategory:UIContentSizeCategoryExtraSmall];
     self.slider.traitCollectionOverride = xsTraitCollection;
+    UIView *valueLabel = [self.slider.thumbTrack valueForKey:@"_valueLabel"];
```
Comment (romoore) [inline components/Slider/tests/snapshot/MDCSliderSnapshotTests.m:260]: Same request for a comment here if these lines are required.

File: components/private/ThumbTrack/src/MDCNumericValueLabel.h [hunk 4]
```diff
@@ -14,18 +14,25 @@
 
 #import <UIKit/UIKit.h>
 
-@interface MDCNumericValueLabel : UIView
+@interface MDCNumericValueLabel : UIView <UIContentSizeCategoryAdjusting>
 
 /** The background color of the value label. */
-@property(nonatomic, retain) UIColor *backgroundColor;
+@property(nonatomic, retain, nonnull) UIColor *backgroundColor;
```
Comment (romoore) [inline components/private/ThumbTrack/src/MDCNumericValueLabel.h]: Tip: Nit: Since you're updating the property semantics, could you switch `retain` with `strong`? I think `retain` is now considered archaic.

File: components/Slider/src/MDCSlider.h [hunk 5]
```diff
@@ -367,6 +368,17 @@ IB_DESIGNABLE
...
+ */
+@property(nonatomic, strong, null_resettable) UIFont *discreteValueLabelFont;
```
Comment (romoore) [inline components/Slider/src/MDCSlider.h:381]: Please add unit tests for both the default value and the effect of setting `nil` (the reset effect).

File: components/private/ThumbTrack/src/MDCThumbTrack.h [hunk 6]
```diff
@@ -226,6 +226,17 @@
...
+/**
+ The font of the discrete value label.
+
+ This font will come into effect only when @c numDiscreteValues is larger than 0 and when @c
+ shouldDisplayDiscreteValueLabel is
+ @c YES.
+
+ Defaults to [[MDCTypography fontLoader] regularFontOfSize:12].
+ */
+@property(nonatomic, strong, null_resettable) UIFont *discreteValueLabelFont;
```
Comment (romoore) [inline components/private/ThumbTrack/src/MDCThumbTrack.h:238]: Please add unit tests for the default value and the null-reset effect.

File: components/private/ThumbTrack/src/MDCThumbTrack.m [hunk 7]
```diff
@@ -822,14 +842,21 @@ - (void)updateViewsMainIsAnimated:(__unused BOOL)animated
     if ([_delegate respondsToSelector:@selector(thumbTrack:stringForValue:)]) {
       _valueLabel.text = [_delegate thumbTrack:self stringForValue:_value];
       if (CGRectGetWidth(_valueLabel.frame) > 1) {
+        CGFloat scale = UIScreen.mainScreen.scale;
```
Comment (romoore) [inline components/private/ThumbTrack/src/MDCThumbTrack.m]: Nit: Please use the view's current screen's scale (if available).  Something like this:

```objc
CGFloat scale = self.window.screen.scale > 0 ?: UIScreen.mainScreen.scale;
```

**Review by romoore — APPROVED** [review]
LGTM as a definite improvement for Slider and ThumbTrack! 

I'm a little worried at the shape/appearance of the value label with large font sizes. However, I suspect it's something that needs designer input rather than anything we can resolve quickly.   I've filed #8714 to track that work.

File: components/Slider/src/MDCSlider.h (hunk 1 above)
Comment (yarneo) [reply components/Slider/src/MDCSlider.h:378]: done

File: components/Slider/tests/snapshot/MDCSliderSnapshotTests.m (hunk 2 above)
Comment (yarneo) [reply components/Slider/tests/snapshot/MDCSliderSnapshotTests.m:224]: added comments

File: components/private/ThumbTrack/src/MDCNumericValueLabel.h (hunk 4 above)
Comment (yarneo) [reply components/private/ThumbTrack/src/MDCNumericValueLabel.h]: done

File: components/Slider/src/MDCSlider.h (hunk 5 above)
Comment (yarneo) [reply components/Slider/src/MDCSlider.h:381]: done

File: components/private/ThumbTrack/src/MDCThumbTrack.h (hunk 6 above)
Comment (yarneo) [reply components/private/ThumbTrack/src/MDCThumbTrack.h:238]: done

File: components/private/ThumbTrack/src/MDCThumbTrack.m (hunk 7 above)
Comment (yarneo) [reply components/private/ThumbTrack/src/MDCThumbTrack.m]: done
